### PR TITLE
feat: transaction coordinator capable of computing and signing a valid bitcoin transaction

### LINKED ISSUE #317: [Feature]: Transaction coordinator capable of loading a functioning coordinator state machine
### Context
In #316 the skeleton implementation of the transaction coordinator was outlined. As a next step, the coordinator should be able to load the appropriate aggregate key and signer set for a signing round, and use this information to construct a WSTS coordinator state machine which it can use to coordinate a signing round.

### Definition of done
The transaction coordinator is capable of loading a functioning coordinator state machine, which together with signers can produce valid signatures for arbitrary messages.

### LINKED ISSUE #319: [Feature]: Transaction coordinator capable of computing and signing a valid bitcoin transaction
Depends on #317 and #318 

### Context
Once the transaction coordinator is able to load a functioning wsts coordinator state machine, we should implement the ability for the coordinator to collect all the necessary information to construct a valid transaction package and provide the ability to run a signing round with this transaction package.

### Definition of done
The transaction coordinator can compute and sign valid bitcoin transactions.

### PR DESCRIPTION
**Pull request by netrome**
closes #319, closes #317 too,

This PR makes the transaction coordinator capable of computing a bitcoin transaction package and having this package signed by a set of signers, along with a unit test which runs this logic together with a WSTS signer set.

### CODE REVIEW

File: signer/src/testing/wsts.rs [hunk 1]
```diff
@@ -409,6 +449,27 @@ impl SignerSet {
...
+        let mut signer_handles = Vec::new();
+        for signer in self.signers.drain(..) {
+            let handle = tokio::spawn(async { signer.run_until_signature_share_response().await });
+            signer_handles.push(handle);
+        }
+
+        for handle in signer_handles {
+            let signer = handle.await.expect("signer crashed");
+            self.signers.push(signer)
+        }
```
Comment (djordon) [inline signer/src/testing/wsts.rs:470]: I think you can use `futures::future::join_all`, or something like that, but without the spawn.

Comment (djordon) [reply signer/src/testing/wsts.rs:470]: Ohh, `run_until_signature_share_response` takes ownership of the full type, and returns it. This is why we use `drain` here. No feedback, just thought that I'd note this.

File: signer/src/testing/wsts.rs [hunk 2]
```diff
@@ -409,6 +449,27 @@ impl SignerSet {
         )
     }
 
+    /// Participate in N signing rounds
+    pub async fn participate_in_signing_rounds_forever(&mut self) {
+        loop {
+            self.participate_in_signing_round().await
+        }
+    }
```
Comment (djordon) [inline signer/src/testing/wsts.rs:457]: Why forever and not N signing rounds?

File: signer/src/testing/transaction_coordinator.rs [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+        let first_script_pubkey = broadcasted_tx
+            .tx_out(0)
+            .expect("missing tx output")
+            .script_pubkey
+            .clone();
+
+        handle.stop_event_loop().await;
+
+        assert_eq!(first_script_pubkey, aggregate_key.signers_script_pubkey());
```
Comment (djordon) [inline signer/src/testing/transaction_coordinator.rs:213]: Awesome 🥇

**Review by djordon — COMMENTED** [review]
🙏🏽 Looks largely good. I think we should make sure none of the prod loops go on for forever.

File: signer/src/transaction_coordinator.rs [hunk 4]
```diff
@@ -207,28 +215,158 @@ where
...
+        bitcoin_chain_tip: &model::BitcoinBlockHash,
+        coordinator_state_machine: &mut wsts_state_machine::CoordinatorStateMachine,
+        txid: bitcoin::Txid,
+    ) -> Result<wsts::taproot::SchnorrProof, error::Error> {
+        loop {
+            let msg = self.network.receive().await?;
+
+            let message::Payload::WstsMessage(wsts_msg) = msg.inner.payload else {
+                continue;
+            };
```
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_coordinator.rs:341]: Yeah I agree. In the ideal case only `WstsMessages` for this particular coordinator should be in-flight when the coordinator attempts to coordinate a signing round, but if we have multiple signers believing they are the coordinator this does become problematic. However, with the suggested timeout I suppose that scenario would just entail this coordinator timing out, restarting and realizing it's not the coordinator anymore so I think we should be fine.

There could be other edge cases as well, so we should take that into consideration. I suppose we'd also like to filter out any messages that doesn't relate to this particular chain tip for example so the "right" coordinator doesn't get confused by seeing messages from stale coordinators trying to coordinate a signing round.

I'll create a follow-up ticket to add the chain tip validation and a test case for it.

File: signer/src/transaction_coordinator.rs [hunk 5]
```diff
@@ -207,28 +215,158 @@ where
...
+    }
+
+    #[tracing::instrument(skip(self))]
+    async fn relay_messages_to_wsts_state_machine_until_signature_created(
+        &mut self,
+        bitcoin_chain_tip: &model::BitcoinBlockHash,
+        coordinator_state_machine: &mut wsts_state_machine::CoordinatorStateMachine,
+        txid: bitcoin::Txid,
+    ) -> Result<wsts::taproot::SchnorrProof, error::Error> {
+        loop {
```
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_coordinator.rs:336]: Agree. I'll add one.

File: signer/src/transaction_coordinator.rs [hunk 6]
```diff
@@ -266,11 +404,23 @@ where
...
+            .await?
+            .ok_or(error::Error::MissingSignerUtxo)?;
+        let last_fees = self.bitcoin_client.get_last_fee(utxo.outpoint).await?;
+
+        Ok(utxo::SignerBtcState {
+            fee_rate,
+            utxo,
+            public_key: bitcoin::XOnlyPublicKey::from(aggregate_key),
+            last_fees,
+            magic_bytes: [0, 0],
```
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_coordinator.rs:422]: Right, thanks for spotting. Will do.

File: signer/src/wsts_state_machine.rs [hunk 7]
```diff
@@ -317,7 +318,7 @@ impl std::ops::DerefMut for CoordinatorStateMachine {
     }
 }
 
-/// Convert an error
+/// Convert a coordinator error to a `error::Error`
```
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/wsts_state_machine.rs:321]: Yes, thanks for noticing!

File: signer/src/wsts_state_machine.rs [hunk 8]
```diff
@@ -269,7 +269,8 @@ impl CoordinatorStateMachine {
         // In this part we are trying to set the party_polynomials of the
         // WstsCoordinator given all of the known public keys that we
         // stored in the database.
-        for msg in public_dkg_shares.values().cloned() {
+        for mut msg in public_dkg_shares.values().cloned() {
+            msg.dkg_id = 1;
```
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/wsts_state_machine.rs:273]: Yes, in this scenario the messages we replay comes from the prevoius dkg round and they will have `dkg_id=0`. Therefore, manually setting them to 1 is necessary to make the state machine recognize them.

File: signer/src/testing/wsts.rs (hunk 1 above)
Comment (netrome) [reply signer/src/testing/wsts.rs:470]: Exactly, we return the signers when we await the handle.

File: signer/src/testing/transaction_coordinator.rs (hunk 3 above)
Comment (netrome) [reply signer/src/testing/transaction_coordinator.rs:213]: Thanks ☺️

File: signer/src/transaction_coordinator.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_coordinator.rs:341]: Created https://github.com/stacks-network/sbtc/issues/465

File: signer/src/testing/wsts.rs (hunk 2 above)
Comment (netrome) [reply signer/src/testing/wsts.rs:457]: Oh the comment is stale. I first implemented it for N signing rounds, only to find that the number of transactions constructed varied between rounds (which is strange, but our tests are unfortunately not completely deterministic).

I did not find a good way to control this number in the test case, so I just wrote the test case to not assume a fixed number instead.

File: signer/src/transaction_coordinator.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (djordon) [reply signer/src/transaction_coordinator.rs:341]: 🙏🏽

File: signer/src/transaction_coordinator.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_coordinator.rs:422]: Todo and follow-up issue created https://github.com/stacks-network/sbtc/issues/472

File: signer/src/wsts_state_machine.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/wsts_state_machine.rs]: Done.

File: signer/src/transaction_coordinator.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_coordinator.rs]: Done.

File: signer/src/transaction_coordinator.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (netrome) [reply signer/src/transaction_coordinator.rs:344]: Done.

File: signer/src/testing/wsts.rs (hunk 2 above)
Comment (netrome) [reply signer/src/testing/wsts.rs]: Updated the comment.

**Review by djordon — APPROVED** [review]
💪🏽 ✅ Looks good!

**Pull request merged**